plugins/hls-haddock-comments-plugin/src/Ide/Plugin/HaddockComments.hs

```diff
@@ -0,0 +1,154 @@
+        anns = relativiseApiAnns <$> (pm_parsed_source <$> pm) <*> (pm_annotations <$> pm)
+        edits = [runGenComments gen locDecls anns range | noErr, gen <- genList]
+    return $ Right $ List [CACodeAction $ toAction title uri edit | (Just (title, edit)) <- edits]
+
+genList :: [GenComments]
+genList =
+  [ genForSig,
+    genForRecord
+  ]
+
+-----------------------------------------------------------------------------
+data GenComments = forall a.
+  GenComments
+  { title :: T.Text,
+    fromDecl :: HsDecl GhcPs -> Maybe a,
+    collectKeys :: a -> [AnnKey],
+    isFresh :: Annotation -> Bool,
+    updateAnn :: Annotation -> Annotation
+  }
+
+runGenComments :: GenComments -> Maybe [LHsDecl GhcPs] -> Maybe Anns -> Range -> Maybe (T.Text, TextEdit)
+runGenComments GenComments {..} mLocDecls mAnns range
+  | Just locDecls <- mLocDecls,
+    Just anns <- mAnns,
+    [(locDecl, src, x)] <- [(locDecl, l, x) | locDecl@(L l (fromDecl -> Just x)) <- locDecls, inRange range l],
+    annKeys <- collectKeys x,
+    not $ null annKeys,
+    and $ maybe False isFresh . flip Map.lookup anns <$> annKeys,
+    anns' <- foldr (Map.adjust updateAnn) anns annKeys,
+    Just range' <- calcRange src range,
+    result <- T.strip . T.pack $ exactPrint locDecl anns' =
+    Just (title, TextEdit range' result)
+  | otherwise = Nothing
+
+-----------------------------------------------------------------------------
+
+genForSig :: GenComments
+genForSig = GenComments {..}
+  where
+    title = "Generate signature comments"
+
+    fromDecl (SigD _ (TypeSig _ _ (HsWC _ (HsIB _ x)))) = Just x
+    fromDecl _ = Nothing
+    updateAnn x = x {annEntryDelta = DP (0, 1), annsDP = dp}
+
+    isFresh Ann {annsDP}
+      | null [() | (AnnComment _, _) <- annsDP] = True
+      | otherwise = False
+
+    collectKeys = keyFromTyVar 0
+
+    comment = mkComment "-- ^ " noSrcSpan
+    dp = [(AnnComment comment, DP (0, 1)), (G AnnRarrow, DP (1, 2))]
+
+genForRecord :: GenComments
+genForRecord = GenComments {..}
+  where
+    title = "Generate fields comments"
+
+    fromDecl (TyClD _ DataDecl {tcdDataDefn = HsDataDefn {dd_cons = cons}}) = Just [x | (L _ ConDeclH98 {con_args = x}) <- cons]
+    fromDecl _ = Nothing
+
+    updateAnn x = x {annEntryDelta = DP (1, -7), annPriorComments = [(comment, DP (1, -7))]}
+
+    isFresh Ann {annPriorComments} = null annPriorComments
+
+    collectKeys = keyFromCon
+
+    comment = mkComment "-- | " noSrcSpan
+
+-----------------------------------------------------------------------------
+
+toAction :: T.Text -> Uri -> TextEdit -> CodeAction
+toAction title uri edit = CodeAction {..}
+  where
+    _title = title
+    _kind = Just CodeActionQuickFix
+    _diagnostics = Nothing
+    _command = Nothing
+    _changes = Just $ HashMap.singleton uri $ List [edit]
+    _documentChanges = Nothing
+    _edit = Just WorkspaceEdit {..}
+
+calcRange :: SrcSpan -> Range -> Maybe Range
+calcRange src range
+  | inRange range src,
```

Comment:
Other than this check, it looks like `range` is not used in this function.
And I believe this function is called only when `inRange range src` is already checked...

Reply:
Ah, you are right :-)